ext/mbstring/mbstring.c

```diff
@@ -4810,8 +4810,88 @@ static bool mb_fast_check_utf8_default(zend_string *str)
 
 	return true;
 # else
-	/* No SSE2 support; we might add generic UTF-8 specific validation code here later */
-	return php_mb_check_encoding(ZSTR_VAL(str), ZSTR_LEN(str), &mbfl_encoding_utf8);
+	/* This UTF-8 validation function is derived from PCRE2 */
+	size_t length = ZSTR_LEN(str);
+	/* Table of the number of extra bytes, indexed by the first byte masked with
+	0x3f. The highest number for a valid UTF-8 first byte is in fact 0x3d. */
+	static const uint8_t utf8_table[] = {
+		1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+		1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+		2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+		3,3,3,3,3,3,3,3,4,4,4,4,5,5,5,5
+	};
+
+	for (; length > 0; p++) {
+		uint32_t d;
+		unsigned char c = *p;
+		length--;
+
+		if (c < 128) {
+			/* ASCII character */
+			continue;
+		}
+
+		if (c < 0xc0) {
+			/* Isolated 10xx xxxx byte */
+			return false;
+		}
+
+		if (c >= 0xfe) {
+			/* Invalid 0xfe or 0xff bytes */
+			return false;
+		}
+
+		uint32_t ab = utf8_table[c & 0x3f]; /* Number of additional bytes (1-5) */
+		if (length < ab) {
+			/* Missing bytes */
+			return false;
+		}
+		length -= ab;
+
+		/* Check top bits in the second byte */
+		if (((d = *(++p)) & 0xc0) != 0x80) {
+			return false;
+		}
+
+		/* For each length, check that the remaining bytes start with the 0x80 bit
+		 * set and not the 0x40 bit. Then check for an overlong sequence, and for the
+		 * excluded range 0xd800 to 0xdfff. */
+		switch (ab) {
+		case 1:
+			/* 2-byte character. No further bytes to check for 0x80. Check first byte
+			 * for for xx00 000x (overlong sequence). */
+			if ((c & 0x3e) == 0) {
+				return false;
+			}
+			break;
+
+		case 2:
+			/* 3-byte character. Check third byte for 0x80. Then check first 2 bytes for
+			 * 1110 0000, xx0x xxxx (overlong sequence) or 1110 1101, 1010 xxxx (0xd800-0xdfff) */
+			if ((*(++p) & 0xc0) != 0x80 || (c == 0xe0 && (d & 0x20) == 0) || (c == 0xed && d >= 0xa0)) {
+				return false;
+			}
+			break;
+
+		case 3:
+			/* 4-byte character. Check 3rd and 4th bytes for 0x80. Then check first 2
+			 * bytes for for 1111 0000, xx00 xxxx (overlong sequence), then check for a
+			 * character greater than 0x0010ffff (f4 8f bf bf) */
+			if ((*(++p) & 0xc0) != 0x80 || (*(++p) & 0xc0) != 0x80 || (c == 0xf0 && (d & 0x30) == 0) || (c > 0xf4 || (c == 0xf4 && d > 0x8f))) {
+				return false;
+			}
+			break;
+
+		case 4:
+		case 5:
+			/* 5-byte character. Check 3rd, 4th, and 5th bytes for 0x80. Then check for
+			1111 1000, xx00 0xxx */
+			return false;
+			break;
+		}
```

Comment:
I don't understand those cases (nor why these values are in the UTF8 table). Is this to detect overlong sequences?

As my understanding, a UTF-8 character is encoded in 4 bytes maximum. So having 4 or 5 *extra* bytes sounds like nonsense. Is it just the comment being copied and pasted from previous cases?

Reply:
@Girgias The original RFC which defined UTF-8 made allowance for 5-byte characters, which were to start with `0b111110`, and 6-byte characters, which were to start with `0b1111110`. A later RFC updated the definition of UTF-8 such that these became illegal.

Since the UTF-8 validation routine for PCRE returns, not just true/false, but a specific error code indicating the type of badness which it found, they wanted to detect that case separately.

But for us there is really no point in doing this. I will simplify the code.

Reply:
maybe worrth to submit into PCRE as well

Reply:
> @Girgias The original RFC which defined UTF-8 made allowance for 5-byte characters, which were to start with `0b111110`, and 6-byte characters, which were to start with `0b1111110`. A later RFC updated the definition of UTF-8 such that these became illegal.
> 
> Since the UTF-8 validation routine for PCRE returns, not just true/false, but a specific error code indicating the type of badness which it found, they wanted to detect that case separately.
> 
> But for us there is really no point in doing this. I will simplify the code.

Huh TIL